sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -910,6 +910,30 @@ case class WithCTE(plan: LogicalPlan, cteDefs: Seq[CTERelationDef]) extends Logi
   def withNewPlan(newPlan: LogicalPlan): WithCTE = {
     withNewChildren(children.init :+ newPlan).asInstanceOf[WithCTE]
   }
+
+  override def doCanonicalize(): LogicalPlan = {
+    val canonicalized = super.doCanonicalize().asInstanceOf[WithCTE]
+    val defIndex = canonicalized.cteDefs.map(_.id).zipWithIndex.toMap
+
+    def canonicalizeCTE(plan: LogicalPlan): LogicalPlan = {
+      plan.resolveOperatorsUpWithPruning(
+        _.containsAnyPattern(CTE, PLAN_EXPRESSION)) {
+        case ref: CTERelationRef =>
+          ref.copy(cteId = defIndex(ref.cteId).toLong)
+
```

Comment:
can we make sure there is no nested `WithCTE`?

Reply:
> can we make sure there is no nested `WithCTE`?

Form the CTESubtitution's code, and it's pr, there won't have nested `WithCTE`.  https://github.com/apache/spark/pull/37751 cc @maryannxue

Reply:
shall we add assert to guarantee this assertion?

Reply:
Done

Reply:
@cloud-fan Found nested WithCTE case, added to `cte.sql` and here change the method to support nested CTE, pls take a look again.